### PR TITLE
feat: Allow to upload directories and allow bulk upload

### PR DESCRIPTION
**Pull request by susnux**
* Part of https://github.com/nextcloud/server/issues/43238
* Design for uploader menu is taken from: https://github.com/nextcloud/server/issues/43728

Added a `Directory` class that allows creating a virtual filesystem.
Meaning the directory contains a tree of child nodes.

Then this tree is uploaded level by level within the uploader (new method `bulkUpload`).

The UploadPicker now also support directories when created with `allow-folders`.

---

image

### CODE REVIEW

File: lib/uploader.ts [hunk 1]
```diff
@@ -163,34 +171,194 @@ export class Uploader {
...
+	 *       const { selected, renamed } = await openConflictPicker(currentPath, conflicts, this.fetchContent(currentPath), { recursive: true })
+	 *       return [...selected, ...renamed]
+	 *     } catch (e) {
+	 *       return false
+	 *     }
+	 *   }
+	 * }
+	 * ```
+	 */
+	bulkUpload(destination: string, files: (File|FileSystemEntry)[], callback?: (nodes: Array<File|Directory>, currentPath: string) => Promise<Array<File|Directory>|false>): PCancelable<Upload[]> {
```
Comment (artonge) [inline lib/uploader.ts:219]: The naming "conflicts" with our bulk upload feature, so I would choose another name.

File: lib/components/UploadPicker.vue [hunk 2]
```diff
@@ -240,54 +300,116 @@ export default Vue.extend({
...
 		 * Trigger file picker
+		 * @param folder Upload folders
 		 */
-		onClick() {
-			this.$refs.input.click()
+		onTriggerPick(folder = false) {
+			const input = this.$refs.input as HTMLInputElement
+			// Setup directory picking if enabled
+			if (this.canUploadFolders) {
+				input.webkitdirectory = !!folder
```
Comment (artonge) [inline lib/components/UploadPicker.vue:309]: Maybe more readable if this is what you cant to achieve?

```suggestion
				input.webkitdirectory = folder !== false
```

File: lib/components/UploadPicker.vue [hunk 3]
```diff
@@ -240,54 +300,116 @@ export default Vue.extend({
...
+					}
+
+					// Hanle invalid path
+					if (await this.showInvalidFileNameDialog(file.name)) {
+						// create a new valid path name
+						let newName = file.name
+						this.forbiddenCharacters.forEach((c) => { newName = newName.replaceAll(c, invalidReplacement) })
+						Object.defineProperty(file, 'name', { value: newName })
+						filesToUpload.push(file)
+					}
```
Comment (artonge) [inline lib/components/UploadPicker.vue:387]: Maybe more readable:

```suggestion
					if (invalid) {
						// Hanle invalid path
						if (await this.showInvalidFileNameDialog(file.name)) {
							// create a new valid path name
							let newName = file.name
							this.forbiddenCharacters.forEach((c) => { newName = newName.replaceAll(c, invalidReplacement) })
							Object.defineProperty(file, 'name', { value: newName })
							filesToUpload.push(file)
						} else {
							continue
						}
					}

					filesToUpload.push(file)

```

File: lib/uploader.ts [hunk 4]
```diff
@@ -163,34 +171,194 @@ export class Uploader {
...
+	/**
+	 * Notify listeners of the upload completion
+	 * @param upload The upload that finished
+	 */
+	private _notifyAll(upload: Upload): void {
+		for (const notifier of this._notifiers) {
+			try {
+				notifier(upload)
+			} catch (error) {
+				// ignore
```
Comment (artonge) [inline lib/uploader.ts:183]: Shouldn't we at least log?

File: lib/uploader.ts (hunk 1 above)
Comment (artonge) [inline lib/uploader.ts:219]: Split in multiple lines?
```suggestion
	bulkUpload(
		destination: string,
		files: (File|FileSystemEntry)[],
		callback?: (nodes: Array<File|Directory>, currentPath: string) => Promise<Array<File|Directory>|false>
	): PCancelable<Upload[]> {
```

File: lib/components/UploadPicker.vue [hunk 5]
```diff
@@ -240,54 +300,116 @@ export default Vue.extend({
...
+					// No invalid characters used on this file, so just continue
+					if (!invalid) {
+						filesToUpload.push(file)
+						continue
+					}
+
+					// Hanle invalid path
+					if (await this.showInvalidFileNameDialog(file.name)) {
+						// create a new valid path name
+						let newName = file.name
```
Comment (artonge) [inline lib/components/UploadPicker.vue:383]: There might be new conflicts after the rename, no?

File: lib/components/UploadPicker.vue [hunk 6]
```diff
@@ -240,54 +300,116 @@ export default Vue.extend({
...
+					.show()
+			})
+		},
+
+		async handleConflicts(nodes: Array<File|Directory>, path: string): Promise<Array<File|Directory>|false> {
+			const invalidReplacement = ['-', '_', ' '].filter((c) => !this.forbiddenCharacters.includes(c))[0] ?? 'x'
+
+			try {
+				const content = path === '' ? this.currentContent : await this.getContent(path).catch(() => [])
+				const conflicts = getConflicts(nodes, content)
```
Comment (artonge) [inline lib/components/UploadPicker.vue:362]: Does `getCOnflicts` also check conflicts on folders and does it recursively? Or maybe the nodes list is flat?

Comment (artonge) [reply lib/components/UploadPicker.vue:362]: Actually what happens when the user upload a folder with a name that is already taken by another folder?
Is the new folder name updated? Or is the new folder merged into the existing one?
If the last option is used, it means that we have to recursively get the folder content to check for conflicts?

Comment (susnux) [reply lib/components/UploadPicker.vue:362]: The conflicts are resolved recursively like your file manager does on your OS.
`getConflicts` only gets conflicts in the current level, then it lets you decide what todo:

- Select the new one (aka "write into")
- Select the old one (aka skip the upload)
- Select both: Rename the new one

File: lib/uploader.ts (hunk 4 above)
Comment (susnux) [reply lib/uploader.ts:183]: This was just refactoring, as the same code was already used in two places (chunked upload + normal upload) and would have been duplicated a third time.
So I did not change it, but of course I can add the logging here :)

File: lib/uploader.ts (hunk 1 above)
Comment (susnux) [reply lib/uploader.ts]: Hm then `batchUpload`?

**Comment by susnux** [comment]
Testing is possible with this server PR: https://github.com/nextcloud/server/pull/45095

Just checkout the server link the upload package and build :)

File: lib/components/UploadPicker.vue (hunk 3 above)
Comment (artonge) [reply lib/components/UploadPicker.vue]: My idea was to have only one `filesToUpload.push(file)`, but honestly feel free to ignore as this is superfluous ;)

**Review by skjnldsv — APPROVED** [review]

**Pull request merged**